### ISSUE TITLE
Reconciliation fails even for valid kiali CR after saving incorrect CR several times

### ISSUE THREAD

**Issue by FilipB** [kiali/kiali#2059]
When I save incorrect kiali CR several times and then save correct CR the reconciliation is still failing complaining about the same error which was fixed in the new CR.

**Versions used**
Kiali-ui
    1.12.0 (229a3f4d2c86d28e8bd00a8ddca971e523b733e1)
Kiali
    v1.12.0-SNAPSHOT (d4696cdaaa0b8c4c598f7220ed937b6f69c3f1a3)
Components
Istio
    1.4.2 
Prometheus
    2.12.0 
Kubernetes
    v1.16.2 

**To Reproduce**
1. Install Istio 1.4 via install-istio-via-istioctl.sh -iee true -de true
2. Install the latest kiali operator and kiali via deploy-kiali-operator.sh with following env vars:
OPERATOR_IMAGE_NAME=quay.io/kiali-qe/kiali-operator
OPERATOR_IMAGE_PULL_POLICY=Always
OPERATOR_IMAGE_VERSION=latest
OPERATOR_INSTALL_KIALI=true
OPERATOR_NAMESPACE=kiali-operator
OPERATOR_SKIP_WAIT=false
OPERATOR_VERSION_LABEL=master
OPERATOR_WATCH_NAMESPACE=kiali-operator
ACCESSIBLE_NAMESPACES=default
AUTH_STRATEGY=openshift
CREDENTIALS_CREATE_SECRET=false
GRAFANA_URL=
KIALI_CR=
KIALI_IMAGE_NAME=quay.io/kiali-qe/kiali
KIALI_IMAGE_PULL_POLICY=Always
KIALI_IMAGE_VERSION=latest
ISTIO_NAMESPACE=
JAEGER_URL=
NAMESPACE=istio-system
SECRET_NAME=kiali
3. check that kiali CR contains:
```
external_services:
    grafana:
      url: ''
    tracing:
      url: ''
```
4. update kiali CR via OCP UI console to incorrect config (grafana password is empty):
```
  external_services:
    grafana:
      auth:
        ca_file: /var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt
        password: 
        type: basic
        use_kiali_token: false
        username: internal
      enabled: true
      in_cluster_url: 'https://grafana.istio-system.svc:3000'
      url: test.grafana
    tracing:
      url: ''
```
5. save and wait for reconciliation to finish (which will fail)
6. update kiali CR via OCP UI console to incorrect config (second time just change url):
```
  external_services:
    grafana:
      auth:
        ca_file: /var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt
        password: null
        type: basic
        use_kiali_token: false
        username: internal
      enabled: true
      in_cluster_url: 'https://grafana.istio-system.svc:3000'
      url: test2.grafana
    tracing:
      url: ''
```
7. update kiali CR via OCP UI console to incorrect config (just change url):
```
  external_services:
    grafana:
      auth:
        ca_file: /var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt
        password: null
        type: basic
        use_kiali_token: false
        username: internal
      enabled: true
      in_cluster_url: 'https://grafana.istio-system.svc:3000'
      url: test3.grafana
    tracing:
      url: ''
```
8. change it back to original via OCP UI console:
```
external_services:
    grafana:
      url: ''
    tracing:
      url: ''
```

**Expected behavior**
Reconciliation should be successful with correct kiali CR but it's still failing complaining about:
The error appears to be in
        '/opt/ansible/roles/kiali-deploy/tasks/main.yml': line 434, column 3,
        but may
        be elsewhere in the file depending on the exact syntax problem.
        The offending line appears to be:
        - name: Prepare environment variable for grafana password

Full exception is attached in file. [error.log](https://github.com/kiali/kiali/files/3983824/error.log)
Note that it's important to save incorrect kiali CR more times to see this issue.

**Comment by jshaughn** [comment]
cc @jmazzitelli

**Comment by jmazzitelli** [comment]
i will attempt to replicate and see what the problem is

**Comment by jmazzitelli** [comment]
> * save and wait for reconciliation to finish (which will fail)

The first time here (step 4 and 5) I did not see it fail. Not sure why the behavior is different with an empty password. But since it doesn't seem related to the grafana settings per se, I'll just do something to get this to fail and see what happens. But on the latest 1.11, this doesn't fail with an empty password.

**Comment by jmazzitelli** [comment]
BTW: in order to REMOVE a piece of yaml, you have to set it to `null` - this is a kubernetes quirk of kubernetes yaml "apply" (the merge functionality). So if you want to remove everything, you can't just delete it entirely from the yaml (because then it will merge your yaml with what already exists and thus keep your old settings).

So if you already have a full "grafana" group of settings, this isn't going to do what you think it will (that is, remove all the settings except for "url")

```
external_services:
    grafana:
      url: ""
```

Instead I believe you have to do something like:

```
external_services:
    grafana:
      auth: null
      enabled: null
      in_cluster_url: null
      url: ""
```

See: https://kubectl.docs.kubernetes.io/pages/app_management/field_merge_semantics.html

Its weird, but I've seen this (and been bitten by it) a few times in the past.

**Comment by jmazzitelli** [comment]
(and FWIW: this "null" thing is causing the bug that results in the "template error" you are seeing - I'm trying to figure out how to solve it - there is a problem here with the conditional statement in the ansible when a field is being removed via this "null")

**Comment by jmazzitelli** [comment]
OK, this appears to be a problem everywhere. To remove items, you use the "null" value. This results in the ansible variable to be of type `NoneType`. This means the variable IS defined, its just defined as a NoneType variable - thus our default values do not take effect. But all of the ansible playbook assumes all (well, virtually all) variables are at least their default value (the playbook doesn't have tasks that look to see if something is defined or not if we expect at least the default value). Because the playbook expects default values at least to be set, this NoneType is causing problems in parts of the playbook that expect a variable of type "string" but instead get a variable of type NoneType.

Need to figure out how to fix this. We need the defaults to take effect if the variable is of type NoneType.

**Comment by FilipB** [comment]
> BTW: in order to REMOVE a piece of yaml, you have to set it to `null` - this is a kubernetes quirk of kubernetes yaml "apply" (the merge functionality). So if you want to remove everything, you can't just delete it entirely from the yaml (because then it will merge your yaml with what already exists and thus keep your old settings).
> 
> So if you already have a full "grafana" group of settings, this isn't going to do what you think it will (that is, remove all the settings except for "url")
> 
> ```
> external_services:
>     grafana:
>       url: ""
> ```
> 
> Instead I believe you have to do something like:
> 
> ```
> external_services:
>     grafana:
>       auth: null
>       enabled: null
>       in_cluster_url: null
>       url: ""
> ```
> 
> See: https://kubectl.docs.kubernetes.io/pages/app_management/field_merge_semantics.html
> 
> Its weird, but I've seen this (and been bitten by it) a few times in the past.

It makes kind of sense when merging new config via cli. I was using UI Console to edit the yaml which is doing it via wysiwyg editor.

**Comment by jmazzitelli** [comment]
> I was using UI Console to edit the yaml which is doing it via wysiwyg editor.

which, under the covers, I'm sure is going through the same merge-patch API as the CLI.

But this did flush out a problem that is throughout the playbook. I bet if you do the same kind of test (setting some fields to "null" to remove them) you'll hit other problems like this.

**Comment by jmazzitelli** [comment]
Let me document a workaround since I don't see a quick code fix on the horizon:

WORKAROUND:

If you need to remove a setting from the CR and you hit problems like this, you must DELETE the entire Kiali CR first, let the operator run the reconcile loop (thus uninstalling Kiali) and then CREATE an entirely new Kiali CR with the fields that you want. This will install Kiali with just the settings you want.

**Comment by jmazzitelli** [comment]
Just going to leave some notes here for me as I debug this and see how Ansible behaves.

If you have an item that is NOT a dict (e.g. the `installation_tag` config setting - that's just a string in our variables), and you set it to "null" in the YAML, that variable does NOT become a NoneType in the context of our Ansible playbook. However, if the variable IS a dict (such as `external_services.grafana.auth` - most of our settings are dictionaries), and an element in that dict is null in the YAML, then the playbook will see it as a NoneType in the dict.

You can test if a variable is a NoneType via `the_var is none`.

**Comment by jmazzitelli** [comment]
i believe the "problem" is in our use of `combine()` when someone wants to remove an item from the yaml using null. Lines like this in our vars/main.yml file:

```
{{ kiali_defaults.auth | combine(auth, recursive=True) }}
```

I think combine() is storing the items as NoneType when we just want them to be removed and replaced with our default values. Not sure how to do that yet.

**Comment by jmazzitelli** [comment]
I have a fix ready to go in PR #2063

**Comment by FilipB** [comment]
Just confirming that the fix is working for me. Thank you.
